### PR TITLE
Add locking interface to blockstore

### PR DESCRIPTION
**Pull request by whyrusleeping**
The addition of a locking interface to the blockstore allows us to
perform atomic operations on the underlying datastore without having to
worry about different operations happening in the background, such as
garbage collection.

Also implement a few merkledag package changes.

License: MIT
Signed-off-by: Jeromy <jeromyj@gmail.com>

### CODE REVIEW

File: merkledag/merkledag.go [hunk 1]
```diff
@@ -110,41 +109,86 @@ func (n *dagService) Remove(nd *Node) error {
...
 
-			// Wait for children to finish
-			<-FetchGraph(ctx, nd, serv)
-		}(l)
+			// keep track of open request, when zero, we're done
+			live += len(keys) - 1
+
+			if live == 0 {
+				return nil
+			}
```
Comment (jbenet) [inline merkledag/merkledag.go]: what is `live` used for? it's incorrect. it's not decremented and will never go below 1.

File: merkledag/merkledag.go [hunk 2]
```diff
@@ -110,41 +109,86 @@ func (n *dagService) Remove(nd *Node) error {
...
+		select {
+		case ks, ok := <-in:
+			if !ok {
+				return
+			}
 
-	return done
+			ng := ds.GetNodes(ctx, ks)
+			for _, g := range ng {
+				go func(g NodeGetter) {
```
Comment (jbenet) [inline merkledag/merkledag.go]: can you define this function at the top? the nesting is hard to follow because scopes overlap.

```go
defer close(out)
get := func(g NodeGetter) { ... }
for { ... } 
```

File: merkledag/merkledag.go [hunk 3]
```diff
@@ -269,3 +313,24 @@ func (np *nodePromise) Get(ctx context.Context) (*Node, error) {
 	}
 	return np.cache, nil
 }
+
+// EnumerateChildren will walk the dag below the given root node and add all
+// unseen children to the passed in set.
+// TODO: parallelize to avoid disk latency perf hits?
```
Comment (jbenet) [inline merkledag/merkledag.go]: yeah i think this is going to be very slow-- where is it used?

File: merkledag/merkledag_test.go [hunk 4]
```diff
@@ -214,3 +214,79 @@ func runBatchFetchTest(t *testing.T, read io.Reader) {
...
+
+	root, err := imp.BuildDagFromReader(read, ds, spl, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	err = FetchGraph(context.TODO(), root, ds)
+	if err != nil {
+		t.Fatal(err)
+	}
```
Comment (jbenet) [inline merkledag/merkledag_test.go]: test that the graph is actually fetched?

```go
func FetchGraph(---) { return nil }
```
would pass the test

File: merkledag/merkledag_test.go [hunk 5]
```diff
@@ -214,3 +214,79 @@ func runBatchFetchTest(t *testing.T, read io.Reader) {
...
+
+	root, err := imp.BuildDagFromReader(read, dservs[0], spl, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	err = FetchGraph(context.TODO(), root, dservs[1])
+	if err != nil {
+		t.Fatal(err)
+	}
```
Comment (jbenet) [inline merkledag/merkledag_test.go]: same here

**Comment by jbenet** [comment]
3 comments above

File: merkledag/merkledag.go (hunk 1 above)
Comment (whyrusleeping) [inline merkledag/merkledag.go]: its correct. i guess i could make it clearer by turning `live += len(keys) - 1` into `live += len(keys)` and `live--`

File: merkledag/merkledag.go (hunk 3 above)
Comment (whyrusleeping) [inline merkledag/merkledag.go]: if and when it becomes a bottleneck for some operation, we can introduce a bunch of additional complexity to add concurrency here. In my tests on a datastore of around 2GB, it took less than 8 seconds to run over the entire set of blocks.

File: merkledag/merkledag_test.go (hunk 4 above)
Comment (whyrusleeping) [inline merkledag/merkledag_test.go]: touche

File: merkledag/merkledag.go (hunk 3 above)
Comment (jbenet) [inline merkledag/merkledag.go]: this is going to be used by more than just GC. 

and remember during GC, no adds can be made.

i'm ok merging this and improving it later, but i'm pretty sure we're going to be annoyed by it.  also, we already have concurrent fetch elsewhere-- i imagine it can be reused here.

File: merkledag/merkledag.go (hunk 1 above)
Comment (jbenet) [inline merkledag/merkledag.go]: oh! yeah that's not clear.

**Comment by jbenet** [comment]
Final changes:

```
RLock -> PinLock
Lock -> GCLock
```

**Comment by whyrusleeping** [comment]
@jbenet this RFM?

**Comment by jbenet** [comment]
I think so (rebase first?)



—
